internal/kldtx/addressbook.go

```diff
@@ -0,0 +1,184 @@
+		hr:                  kldutils.NewHTTPRequester("Addressbook", &conf.HTTPRequesterConf),
+		addrToHost:          make(map[string]string),
+		hostToRPC:           make(map[string]kldeth.RPCClientAll),
+	}
+	propNames := &conf.PropNames
+	if propNames.RPCEndpoint == "" {
+		propNames.RPCEndpoint = defaultRPCEndpointProp
+	}
+	if ab.conf.AddressbookURLPrefix != "" && !strings.HasSuffix(ab.conf.AddressbookURLPrefix, "/") {
+		ab.conf.AddressbookURLPrefix += "/"
+	}
+	return ab
+}
+
+type addressBook struct {
+	conf                *AddressBookConf
+	hr                  *kldutils.HTTPRequester
+	mtx                 sync.Mutex
+	fallbackRPCEndpoint string
+	addrToHost          map[string]string
+	hostToRPC           map[string]kldeth.RPCClientAll
+}
+
+// testRPC uses a simple net_version JSON/RPC call to test the health of a cached connection
+func (ab *addressBook) testRPC(rpc kldeth.RPCClient) bool {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	var netID string
+	if err := rpc.CallContext(ctx, &netID, "net_version"); err != nil {
+		log.Errorf("Stale RPC connection in pool: %s", err)
+		return false
+	}
+	return true
+}
+
+func (ab *addressBook) resolveHost(endpoint string) (*url.URL, error) {
+	// We need to connect, but first we need to resolve the hostname to an IP
+	url, err := url.Parse(endpoint)
+	if err != nil {
+		log.Errorf("Invalid URL '%s': %s", endpoint, err)
+		return nil, fmt.Errorf("Invalid URL obtained for address")
+	}
+
+	if ab.conf.HostsFile != "" {
+		hostsMap, err := kldutils.ParseHosts(ab.conf.HostsFile)
+		if err != nil {
+			log.Errorf("Failed to parse hosts file: %s", err)
+			return nil, fmt.Errorf("Configuration problem (hosts file)")
+		}
+		splitHostPort := strings.Split(url.Host, ":")
+		if mappedHost, ok := hostsMap[splitHostPort[0]]; ok && mappedHost != "" {
+			log.Debugf("Resolved %s to %s", url.Host, mappedHost)
+			url.Host = mappedHost
+			if len(splitHostPort) > 1 {
+				url.Host += ":" + splitHostPort[1]
+			}
+		}
+	}
+	return url, nil
+}
+
+// mapEndpoint takes an RPC connect endpoint (prior to host resolution) and maps
+// it to a cached RPC connection. Or creates a new connection and caches it.
+func (ab *addressBook) mapEndpoint(endpoint string) (kldeth.RPCClient, error) {
+
+	// Simple locking on our cache for now (covers long-lived async test+connect operations)
+	ab.mtx.Lock()
+	defer ab.mtx.Unlock()
+
+	// Hopefully we already have a client in our map, and it's healthy
+	rpc, ok := ab.hostToRPC[endpoint]
+	if ok {
+		if ab.testRPC(rpc) {
+			log.Infof("Using cached RPC connection for signing")
+			return rpc, nil
+		}
+		// Clean up our cache entry as it is stale
+		rpc.Close()
+		delete(ab.hostToRPC, endpoint)
+	}
+
+	url, err := ab.resolveHost(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// Connect and cache the RPC connection
+	rpc, err = kldeth.RPCConnect(&kldeth.RPCConnOpts{
+		URL: url.String(),
+	})
+	if err == nil {
```

Comment:
I think we should check `err != nil` and `return nil, err` here:

```
if err != nil {
  return nil, err
}
ab.hostToRPC[endpoint] = rpc
return rpc, nil
```
In the current form, we return the rpc connection even if there is an error